### DISCUSSION TITLE
Initialisation from array doesn't work.

### DISCUSSION THREAD

**Discussion by paul-crashley** [spatie/laravel-data#178] [Q&A]
Just trying this package out for the first time and attempted to create a very simple data object using the `from` array method.

My object:

```
class Object extends Data
{
    public function __construct(
        public int $id,
    ) {
    }
}
```

My initialisation:

```
Object::from([
    'id' => 1,
])
```

But I get the following message:

```
Unresolvable dependency resolving [Parameter #0 [ <required> array $config ]] in class Spatie\LaravelData\Support\DataConfig
```

What am I missing?

**Comment by rubenvanassche** [comment]
Did you publish the config file correctly (normally it should be loaded by default). The `DataConfig` cannot be constructed because the config variable is missing. Also please check that the `DataConfig` is correctly bound into the container by `LaravelDataServiceProvider`

**Reply by telkins** [reply]
I am experiencing the same thing.  I am looking at the docs on installation/setup and it says that the config is optional and there doesn't seem to be anything about binding it into the container.  😕 

I should note that I'm doing this as a part of a package...also using Spatie's package skeleton, fwiw. 🤓

**Reply by telkins** [reply]
OK.  I found where I needed to explicitly add the service provider to the test case.  I've moved past the first hurdle.  🤓

**Reply by edalzell** [reply]
> OK. I found where I needed to explicitly add the service provider to the test case. I've moved past the first hurdle. 🤓

Did you get this solved @telkins I'm getting the same thing and my config file is  published.

**Reply by elabuwa** [reply]
@telkins , can you please add the steps on how you resolved this please? I'm getting stuck as well. My config file is published as well. I get this error when testing though

**Reply by adriangonzales** [reply]
@edalzell @elabuwa 

If you haven't figured this out already, the Data package requires a `$config` object to be provided to the `DataConfig` class. Adding a provider to the TestCase worked for me.

Depending on your needs, you might need to add additional information to the `give([])` to match configuration data the package needs, but this at stopped the error and allowed basic operation of Data objects.

```php
<?php

use Orchestra\Testbench\TestCase as BaseTestCase;
use Spatie\LaravelData\Support\DataConfig;

class TestCase extends BaseTestCase
{
    protected function setUp(): void
    {
        parent::setUp();

        // Provide a config array to DataConfig
        $this->app->when(DataConfig::class)
            ->needs('$config')
            ->give([]);
    }
}
```

**Reply by elabuwa** [reply]
Thank you @adriangonzales .
I ended up extending from "Tests\TestCase" rather than "PHPUnit\Framework\TestCase" for the tests which resolved it for me.

**Reply by telkins** [reply]
> > OK. I found where I needed to explicitly add the service provider to the test case. I've moved past the first hurdle. 🤓
> 
> Did you get this solved @telkins I'm getting the same thing and my config file is published.

I missed this conversation's notifications, so my apologies for the late reply.

The package I was working on was for more of a hobby project, and I haven't been able to devote time to it.  I can try to take a look this weekend to see if I can find what I did that was able to help me move "past the first hurdle"....and then share it.

For now, though, based on what I wrote, I would guess that it had something to do with adding the package service provider explicitly to the test setup.  Again, this was for a package...not a normal laravel app.

It looks like @adriangonzales provides a nice way to do this. 🤓

**Reply by xuandung38** [reply]
> @edalzell @elabuwa
> 
> If you haven't figured this out already, the Data package requires a `$config` object to be provided to the `DataConfig` class. Adding a provider to the TestCase worked for me.
> 
> Depending on your needs, you might need to add additional information to the `give([])` to match configuration data the package needs, but this at stopped the error and allowed basic operation of Data objects.
> 
> ```
> <?php
> 
> use Orchestra\Testbench\TestCase as BaseTestCase;
> use Spatie\LaravelData\Support\DataConfig;
> 
> class TestCase extends BaseTestCase
> {
>     protected function setUp(): void
>     {
>         parent::setUp();
> 
>         // Provide a config array to DataConfig
>         $this->app->when(DataConfig::class)
>             ->needs('$config')
>             ->give([]);
>     }
> }
> ```

Actually, this solution does not solve the problem; it will lead to a new issue 'Error: Call to a member function when() on null'.

**Reply by SaeedNikmehr** [reply]
This error happened to me when I did not put parent::setUp() in my setUp method. after doing that, it worked
